File: JavaTest/src/main/java/test/StringTest.java
package test;

import org.apache.logging.log4j.LogManager;
import org.apache.logging.log4j.Logger;

/**
 * The {@code StringTest} class performs automatic testing of the {@code String}
 * class.
 *
 * @author Petr Kozler
 */
public class StringTest {

    public static Logger logger = LogManager.getLogger(StringTest.class.getName());

    public static void runTestSuite() {
        StringTest.stringTest();
        StringTest.stringTest2();
        StringTest.stringTest3();
        StringTest.compareToTest();
        StringTest.equalsTest();
        StringTest.substringTest();
        StringTest.substringTest2();
        StringTest.indexOfTest();
        StringTest.indexOfTest2();
        StringTest.indexOfTest3();
        StringTest.indexOfTest4();
        StringTest.lengthTest();
        StringTest.trimTest();
        StringTest.toLowerCaseTest();
        StringTest.toUpperCaseTest();
        StringTest.charAtTest();
        StringTest.replaceTest();
        StringTest.startsWithTest();
        StringTest.endsWithTest();
        StringTest.isEmptyTest();
        StringTest.toStringTest();
    }

    /**
     * Tests the constructor with standard string as a parameter.
     */
    public static void stringTest() {
        String a;
        // a string with whitespaces
        a = "This is a test.";
        logger.info(a);
        // letters
        a = "test";
        logger.info(a);
        // digits
        a = "1234";
        logger.info(a);
        // one char
        a = "a";
        logger.info(a);
        // empty string
        a = "";
        logger.info(a);
    }

    /**
     * Tests the constructor with array of bytes as a parameter.
     */
    public static void stringTest2() {
        byte[] a;
        String b;
        // array with some whitespaces
        a = new byte[]{'T', 'h', 'i', 's', ' ', 'i', 's', ' ', 'a', ' ', 't', 'e', 's', 't', '.'};
        b = new String(a);
        logger.info(b);
        // array without whitespaces
        a = new byte[]{'1', '2', '3', '4'};
        b = new String(a);
        logger.info(b);
        // empty array
        a = new byte[]{};
        b = new String(a);
        logger.info(b);
    }

    /**
     * Tests the constructor with array of bytes and indexes as parameters.
     */
    public static void stringTest3() {
        byte[] a;
        String b;
        // choosing a part of array with some chars
        a = new byte[]{'T', 'h', 'i', 's', ' ', 'i', 's', ' ', 'a', ' ', 't', 'e', 's', 't', '.'};
        b = new String(a, 10, 4);
        logger.info(b);
    }

    /**
     * Tests the compareTo method.
     */
    public static void compareToTest() {
        String a, b;
        int c;
        // differens strings at the beginning
        a = "aaa";
        b = "bbb";
        c = a.compareTo(b);
        logger.info(c);
        // different strings at the end
        a = "aaaa";
        b = "aaab";
        c = a.compareTo(b);
        logger.info(c);
        // differens strings at the beginning (swapped)
        a = "bbb";
        b = "aaa";
        c = a.compareTo(b);
        logger.info(c);
        // different strings at the end (swapped)
        a = "aaab";
        b = "aaaa";
        c = a.compareTo(b);
        logger.info(c);
        // equal strings
        a = "aaa";
        b = "aaa";
        c = a.compareTo(b);
        logger.info(c);
    }

    /**
     * Tests the equals method.
     */
    public static void equalsTest() {
        String a, b;
        boolean c;
        a = "aaa";
        b = "bbb";
        c = a.equals(b);
        logger.info(c);
        a = "aaaa";
        b = "aaab";
        c = a.equals(b);
        logger.info(c);
        a = "bbb";
        b = "aaa";
        c = a.equals(b);
        logger.info(c);
        a = "aaab";
        b = "aaaa";
        c = a.equals(b);
        logger.info(c);
        a = "aaa";
        b = "aaa";
        c = a.equals(b);
        logger.info(c);
    }

    /**
     * Tests the substring method.
     */
    public static void substringTest() {
        String a, b;
        a = "This is a test.";
        // string from the beginning
        b = a.substring(0);
        logger.info(b);
        // string from some other position
        b = a.substring(10);
        logger.info(b);
    }

    /**
     * Tests the substring method with end index as a parameter.
     */
    public static void substringTest2() {
        String a, b;
        a = "This is a test.";
        // string from the beginning to the position near the beginning
        b = a.substring(0, 4);
        logger.info(b);
        // string from the beginning to the position far from the beginning
        b = a.substring(10, 14);
        logger.info(b);
        // string from the end
        b = a.substring(14, 15);
        logger.info(b);
        // equal indexes
        b = a.substring(10, 10);
        logger.info(b);
    }

    /**
     * Tests the indexOf method with a char value.
     */
    public static void indexOfTest() {
        String a;
        int b;
        a = "This is a test.";
        // char occurring once
        b = a.indexOf('a');
        logger.info(b);
        // char occuring more times
        b = a.indexOf('t');
        logger.info(b);
        // char occurring once (at the beginning)
        b = a.indexOf('T');
        logger.info(b);
        // char occurring once (at the end)
        b = a.indexOf('.');
        logger.info(b);
        // char not occurring
        b = a.indexOf('_');
        logger.info(b);
    }

    /**
     * Tests the indexOf method with a char value and end index as a parameter.
     */
    public static void indexOfTest2() {
        String a;
        int b;
        a = "This is a test.";
        b = a.indexOf('a', 10);
        logger.info(b);
        b = a.indexOf('t', 10);
        logger.info(b);
        b = a.indexOf('i', 0);
        logger.info(b);
        b = a.indexOf('i', 4);
        logger.info(b);
        b = a.indexOf('i', 10);
        logger.info(b);
    }

    /**
     * Tests the indexOf method with a string.
     */
    public static void indexOfTest3() {
        String a;
        int b;
        a = "This is a test.";
        // occurring string (at the beginning)
        b = a.indexOf("This");
        logger.info(b);
        // occurring string (at the end)
        b = a.indexOf("test");
        logger.info(b);
        // non-occurring string
        b = a.indexOf("aaa");
        logger.info(b);
    }

    /**
     * Tests the indexOf method with a string and end index as a parameter.
     */
    public static void indexOfTest4() {
        String a;
        int b;
        a = "This is a test.";
        b = a.indexOf("This", 0);
        logger.info(b);
        b = a.indexOf("test", 4);
        logger.info(b);
        b = a.indexOf("This", 10);
        logger.info(b);
    }

    /**
     * Tests the length method.
     */
    public static void lengthTest() {
        String a;
        int b;
        // string with multiple chars
        a = "This is a test.";
        b = a.length();
        logger.info(b);
        // string with one char
        a = ".";
        b = a.length();
        logger.info(b);
        // empty string
        a = "";
        b = a.length();
        logger.info(b);
    }

    /**
     * Tests the trim method.
     */
    public static void trimTest() {
        String a, b;
        // string without whitespaces
        a = "test";
        b = a.trim();
        logger.info(b);
        // whitespaces on the left side
        a = "   test";
        b = a.trim();
        logger.info(b);
        // whitespaces on the right side
        a = "test   ";
        b = a.trim();
        logger.info(b);
        // whitespaces on bots sides
        a = " test ";
        b = a.trim();
        logger.info(b);
        // whitespaces inside
        a = "t e s t";
        b = a.trim();
        logger.info(b);
    }

    /**
     * Tests the toLower method.
     */
    public static void toLowerCaseTest() {
        String a, b;
        // all letters lowercase
        a = "test";
        b = a.toLowerCase();
        logger.info(b);
        // all letters uppercase
        a = "TEST";
        b = a.toLowerCase();
        logger.info(b);
        // different cases
        a = "TesT";
        b = a.toLowerCase();
        logger.info(b);
    }

    /**
     * Tests the toUpper method.
     */
    public static void toUpperCaseTest() {
        String a, b;
        a = "test";
        b = a.toUpperCase();
        logger.info(b);
        a = "TEST";
        b = a.toUpperCase();
        logger.info(b);
        a = "TesT";
        b = a.toUpperCase();
        logger.info(b);
    }

    /**
     * Tests the charAt method.
     */
    public static void charAtTest() {
        String a;
        char b;
        a = "This is a test.";
        // at the beginning
        b = a.charAt(0);
        logger.info(b);
        // in the middle
        b = a.charAt(10);
        logger.info(b);
        // at the end
        b = a.charAt(14);
        logger.info(b);
    }

    /**
     * Tests the replace method.
     */
    public static void replaceTest() {
        String a, b;
        a = "This is a test.";
        // replacing chars inside the string
        b = a.replace(' ', '_');
        logger.info(b);
        // replacing chars in the beginning of the string
        b = a.replace('T', 't');
        logger.info(b);
        // replacing chars in the end of the string
        b = a.replace('.', '!');
        logger.info(b);
        // replacing of the non-existent chars
        b = a.replace('?', ':');
        logger.info(b);
    }

    /**
     * Tests the startsWith method.
     */
    public static void startsWithTest() {
        String a, b;
        boolean c;
        a = "This is a test.";
        // string at the beginning
        b = "This ";
        c = a.startsWith(b);
        logger.info(c);
        // string inside
        b = "is";
        c = a.startsWith(b);
        logger.info(c);
        // string at the end
        b = "test.";
        c = a.startsWith(b);
        logger.info(c);
    }

    /**
     * Tests the endsWith method.
     */
    public static void endsWithTest() {
        String a, b;
        boolean c;
        a = "This is a test.";
        b = "This ";
        c = a.endsWith(b);
        logger.info(c);
        b = "is";
        c = a.endsWith(b);
        logger.info(c);
        b = "test.";
        c = a.endsWith(b);
        logger.info(c);
    }

    /**
     * Tests the isEmpty method.
     */
    public static void isEmptyTest() {
        String a;
        boolean b;
        // non-empty string
        a = "This is a test.";
        b = a.isEmpty();
        logger.info(b);
        // non-empty string with a single char
        a = ".";
        b = a.isEmpty();
        logger.info(b);
        // empty string
        a = "";
        b = a.isEmpty();
        logger.info(b);
    }

    /**
     * Tests the toString method.
     */
    public static void toStringTest() {
        String a, b;
        // multiple chars
        a = "This is a test.";
        b = a.toString();
        logger.info(b);
        // one char
        a = ".";
        b = a.toString();
        logger.info(b);
        // no char
        a = "";
        b = a.toString();
        logger.info(b);
    }

}

File: C/CLibrary/tests/StringTest.c
#include <stdio.h>
#include <stdlib.h>

#include "test.h"
#include "String.h"

/**
 * A String module test suite.
 *
 * @author Petr Kozler (A13B0359P)
 */

/**
 * Tests the constructor with standard string as a parameter.
 */

void stringTest() {
    String *a;
    // a string with whitespaces
    a = new_String("This is a test.");
    assertEqualsStr("This is a test.", a->s);
    // letters
    a = new_String("test");
    assertEqualsStr("test", a->s);
    // digits
    a = new_String("1234");
    assertEqualsStr("1234", a->s);
    // one char
    a = new_String("a");
    assertEqualsStr("a", a->s);
    // empty string
    a = new_String("");
    assertEqualsStr("", a->s);
}

/**
 * Tests the constructor with array of bytes as a parameter.
 */

void stringTest2() {
    const int32_t length = 15;
    const int32_t length2 = 4;
    const int32_t length3 = 0;
    int8_t *a;
    String *b;
    // array with some whitespaces
    a = malloc(length * sizeof(int8_t));
    a[0] = (int8_t) 'T';
    a[1] = (int8_t) 'h';
    a[2] = (int8_t) 'i';
    a[3] = (int8_t) 's';
    a[4] = (int8_t) ' ';
    a[5] = (int8_t) 'i';
    a[6] = (int8_t) 's';
    a[7] = (int8_t) ' ';
    a[8] = (int8_t) 'a';
    a[9] = (int8_t) ' ';
    a[10] = (int8_t) 't';
    a[11] = (int8_t) 'e';
    a[12] = (int8_t) 's';
    a[13] = (int8_t) 't';
    a[14] = (int8_t) '.';
    b = new_StringB(a, length);
    assertEqualsStr("This is a test.", b->s);
    // array without whitespaces
    a = malloc(length2 * sizeof(int8_t));
    a[0] = (int8_t) '1';
    a[1] = (int8_t) '2';
    a[2] = (int8_t) '3';
    a[3] = (int8_t) '4';
    b = new_StringB(a, length2);
    assertEqualsStr("1234", b->s);
    // empty array
    a = malloc(length3 * sizeof(int8_t));
    b = new_StringB(a, length3);
    assertEqualsStr("", b->s);
}

/**
 * Tests the constructor with array of bytes and indexes as parameters.
 */

void stringTest3() {
    const int32_t length = 15;
    int8_t *a;
    String *b;
    // choosing a part of array with some chars
    a = malloc(length * sizeof(int8_t));
    a[0] = (int8_t) 'T';
    a[1] = (int8_t) 'h';
    a[2] = (int8_t) 'i';
    a[3] = (int8_t) 's';
    a[4] = (int8_t) ' ';
    a[5] = (int8_t) 'i';
    a[6] = (int8_t) 's';
    a[7] = (int8_t) ' ';
    a[8] = (int8_t) 'a';
    a[9] = (int8_t) ' ';
    a[10] = (int8_t) 't';
    a[11] = (int8_t) 'e';
    a[12] = (int8_t) 's';
    a[13] = (int8_t) 't';
    a[14] = (int8_t) '.';
    b = new_StringRangeB(a, 10, 4);
    assertEqualsStr("test", b->s);
}

/**
 * Tests the compareTo function.
 */

void compareToTest() {
    String *a, *b;
    int32_t c;
    // differens strings at the beginning
    a = new_String("aaa");
    b = new_String("bbb");
    c = compareToStr(a, b);
    assertEqualsI(-1, c);
    // different strings at the end
    a = new_String("aaaa");
    b = new_String("aaab");
    c = compareToStr(a, b);
    assertEqualsI(-1, c);
    // differens strings at the beginning (swapped)
    a = new_String("bbb");
    b = new_String("aaa");
    c = compareToStr(a, b);
    assertEqualsI(1, c);
    // different strings at the end (swapped)
    a = new_String("aaab");
    b = new_String("aaaa");
    c = compareToStr(a, b);
    assertEqualsI(1, c);
    // equal strings
    a = new_String("aaa");
    b = new_String("aaa");
    c = compareToStr(a, b);
    assertEqualsI(0, c);
}

/**
 * Tests the equals function.
 */

void equalsTest() {
    String *a, *b;
    bool c;
    a = new_String("aaa");
    b = new_String("bbb");
    c = equalsStr(a, b);
    assertEqualsBool(false, c);
    a = new_String("aaaa");
    b = new_String("aaab");
    c = equalsStr(a, b);
    assertEqualsBool(false, c);
    a = new_String("bbb");
    b = new_String("aaa");
    c = equalsStr(a, b);
    assertEqualsBool(false, c);
    a = new_String("aaab");
    b = new_String("aaaa");
    c = equalsStr(a, b);
    assertEqualsBool(false, c);
    a = new_String("aaa");
    b = new_String("aaa");
    c = equalsStr(a, b);
    assertEqualsBool(true, c);
}

/**
 * Tests the substring function.
 */

void substringTest() {
    String *a, *b;
    a = new_String("This is a test.");
    // string from the beginning
    b = substring(a, 0);
    assertEqualsStr("This is a test.", b->s);
    // string from some other position
    b = substring(a, 10);
    assertEqualsStr("test.", b->s);
}

/**
 * Tests the substring function with end index as a parameter.
 */

void substringTest2() {
    String *a, *b;
    a = new_String("This is a test.");
    // string from the beginning to the position near the beginning
    b = substringTo(a, 0, 4);
    assertEqualsStr("This", b->s);
    // string from the beginning to the position far from the beginning
    b = substringTo(a, 10, 14);
    assertEqualsStr("test", b->s);
    // string from the end
    b = substringTo(a, 14, 15);
    assertEqualsStr(".", b->s);
    // equal indexes
    b = substringTo(a, 10, 10);
    assertEqualsStr("", b->s);
}

/**
 * Tests the indexOf function with a char value.
 */

void indexOfTest() {
    String *a;
    int32_t b;
    a = new_String("This is a test.");
    // char occurring once
    b = indexOfC(a, 'a');
    assertEqualsI(8, b);
    // char occuring more times
    b = indexOfC(a, 't');
    assertEqualsI(10, b);
    // char occurring once (at the beginning)
    b = indexOfC(a, 'T');
    assertEqualsI(0, b);
    // char occurring once (at the end)
    b = indexOfC(a, '.');
    assertEqualsI(14, b);
    // char not occurring
    b = indexOfC(a, '_');
    assertEqualsI(-1, b);
}

/**
 * Tests the indexOf function with a char value and end index as a parameter.
 */

void indexOfTest2() {
    String *a;
    int32_t b;
    a = new_String("This is a test.");
    b = indexOfFromC(a, 'a', 10);
    assertEqualsI(-1, b);
    b = indexOfFromC(a, 't', 10);
    assertEqualsI(10, b);
    b = indexOfFromC(a, 'i', 0);
    assertEqualsI(2, b);
    b = indexOfFromC(a, 'i', 4);
    assertEqualsI(5, b);
    b = indexOfFromC(a, 'i', 10);
    assertEqualsI(-1, b);
}

/**
 * Tests the indexOf function with a string.
 */

void indexOfTest3() {
    String *a;
    int32_t b;
    a = new_String("This is a test.");
    // occurring string (at the beginning)
    b = indexOfStr(a, new_String("This"));
    assertEqualsI(0, b);
    // occurring string (at the end)
    b = indexOfStr(a, new_String("test"));
    assertEqualsI(10, b);
    // non-occurring string
    b = indexOfStr(a, new_String("aaa"));
    assertEqualsI(-1, b);
}

/**
 * Tests the indexOf function with a string and end index as a parameter.
 */

void indexOfTest4() {
    String *a;
    int32_t b;
    a = new_String("This is a test.");
    b = indexOfFromStr(a, new_String("This"), 0);
    assertEqualsI(0, b);
    b = indexOfFromStr(a, new_String("test"), 4);
    assertEqualsI(10, b);
    b = indexOfFromStr(a, new_String("This"), 10);
    assertEqualsI(-1, b);
}

/**
 * Tests the length function.
 */

void lengthTest() {
    String *a;
    int32_t b;
    // string with multiple chars
    a = new_String("This is a test.");
    b = length(a);
    assertEqualsI(15, b);
    // string with one char
    a = new_String(".");
    b = length(a);
    assertEqualsI(1, b);
    // empty string
    a = new_String("");
    b = length(a);
    assertEqualsI(0, b);
}

/**
 * Tests the trim function.
 */

void trimTest() {
    String *a, *b;
    // string without whitespaces
    a = new_String("test");
    b = trim(a);
    assertEqualsStr("test", b->s);
    // whitespaces on the left side
    a = new_String("   test");
    b = trim(a);
    assertEqualsStr("test", b->s);
    // whitespaces on the right side
    a = new_String("test   ");
    b = trim(a);
    assertEqualsStr("test", b->s);
    // whitespaces on bots sides
    a = new_String(" test ");
    b = trim(a);
    assertEqualsStr("test", b->s);
    // whitespaces inside
    a = new_String("t e s t");
    b = trim(a);
    assertEqualsStr("t e s t", b->s);
}

/**
 * Tests the toLower function.
 */

void toLowerCaseTest() {
    String *a, *b;
    // all letters lowercase
    a = new_String("test");
    b = toLowerCase(a);
    assertEqualsStr("test", b->s);
    // all letters uppercase
    a = new_String("TEST");
    b = toLowerCase(a);
    assertEqualsStr("test", b->s);
    // different cases
    a = new_String("TesT");
    b = toLowerCase(a);
    assertEqualsStr("test", b->s);
}

/**
 * Tests the toUpper function.
 */

void toUpperCaseTest() {
    String *a, *b;
    a = new_String("test");
    b = toUpperCase(a);
    assertEqualsStr("TEST", b->s);
    a = new_String("TEST");
    b = toUpperCase(a);
    assertEqualsStr("TEST", b->s);
    a = new_String("TesT");
    b = toUpperCase(a);
    assertEqualsStr("TEST", b->s);
}

/**
 * Tests the charAt function.
 */

void charAtTest() {
    String *a;
    char b;
    a = new_String("This is a test.");
    // at the beginning
    b = charAt(a, 0);
    assertEqualsC('T', b);
    // in the middle
    b = charAt(a, 10);
    assertEqualsC('t', b);
    // at the end
    b = charAt(a, 14);
    assertEqualsC('.', b);
}

/**
 * Tests the replace function.
 */

void replaceTest() {
    String *a, *b;
    a = new_String("This is a test.");
    // replacing chars inside the string
    b = replace(a, ' ', '_');
    assertEqualsStr("This_is_a_test.", b->s);
    // replacing chars in the beginning of the string
    b = replace(a, 'T', 't');
    assertEqualsStr("this is a test.", b->s);
    // replacing chars in the end of the string
    b = replace(a, '.', '!');
    assertEqualsStr("This is a test!", b->s);
    // replacing of the non-existent chars
    b = replace(a, '?', ':');
    assertEqualsStr("This is a test.", b->s);
}

/**
 * Tests the startsWith function.
 */

void startsWithTest() {
    String *a, *b;
    bool c;
    a = new_String("This is a test.");
    // string at the beginning
    b = new_String("This ");
    c = startsWith(a, b);
    assertEqualsBool(true, c);
    // string inside
    b = new_String("is");
    c = startsWith(a, b);
    assertEqualsBool(false, c);
    // string at the end
    b = new_String("test.");
    c = startsWith(a, b);
    assertEqualsBool(false, c);
}

/**
 * Tests the endsWith function.
 */

void endsWithTest() {
    String *a, *b;
    bool c;
    a = new_String("This is a test.");
    b = new_String("This ");
    c = endsWith(a, b);
    assertEqualsBool(false, c);
    b = new_String("is");
    c = endsWith(a, b);
    assertEqualsBool(false, c);
    b = new_String("test.");
    c = endsWith(a, b);
    assertEqualsBool(true, c);
}

/**
 * Tests the isEmpty function.
 */

void isEmptyTest() {
    String *a;
    bool b;
    // non-empty string
    a = new_String("This is a test.");
    b = isEmptyStr(a);
    assertEqualsBool(false, b);
    // non-empty string with a single char
    a = new_String(".");
    b = isEmptyStr(a);
    assertEqualsBool(false, b);
    // empty string
    a = new_String("");
    b = isEmptyStr(a);
    assertEqualsBool(true, b);
}

int main(int argc, char** argv) {
    const int testCount = 20;
    setUpTestModule("StringTest", testCount);

    registerTest(stringTest, "stringTest");
    registerTest(stringTest2, "stringTest2");
    registerTest(stringTest3, "stringTest3");
    registerTest(compareToTest, "compareToTest");
    registerTest(equalsTest, "equalsTest");
    registerTest(substringTest, "substringTest");
    registerTest(substringTest2, "substringTest2");
    registerTest(indexOfTest, "indexOfTest");
    registerTest(indexOfTest2, "indexOfTest2");
    registerTest(indexOfTest3, "indexOfTest3");
    registerTest(indexOfTest4, "indexOfTest4");
    registerTest(lengthTest, "lengthTest");
    registerTest(trimTest, "trimTest");
    registerTest(toLowerCaseTest, "toLowerCaseTest");
    registerTest(toUpperCaseTest, "toUpperCaseTest");
    registerTest(charAtTest, "charAtTest");
    registerTest(replaceTest, "replaceTest");
    registerTest(startsWithTest, "startsWithTest");
    registerTest(endsWithTest, "endsWithTest");
    registerTest(isEmptyTest, "isEmptyTest");
    
    runTests();

    return (EXIT_SUCCESS);
}
